File: outgoing-simple-java-demo/src/main/java/de/codecentric/cm4sb/demo/outgoing/endpoint/MovieController.java
package de.codecentric.cm4sb.demo.outgoing.endpoint;

import de.codecentric.cm4sb.demo.outgoing.domain.Movie;
import de.codecentric.cm4sb.demo.outgoing.domain.MovieWithSuccessor;
import de.codecentric.cm4sb.demo.outgoing.service.MovieService;
import de.codecentric.cm4sb.demo.outgoing.service.MovieSuccessorService;

import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class MovieController {

  private final MovieService movieService;
  private final MovieSuccessorService movieSuccessorService;

  public MovieController(
          MovieService movieService,
          MovieSuccessorService movieSuccessorService){
    this.movieService = movieService;
    this.movieSuccessorService = movieSuccessorService;
  }

  @GetMapping(path = "/movies", produces = MediaType.APPLICATION_JSON_VALUE)
  public MovieWithSuccessor getRecommendation(@RequestParam(required = false) String title){
    Movie movie = getRandomMovieOrByTitle(title);
    System.out.println("Recommend movie: $movie");
    var movieSuccessor = movieSuccessorService.getMovieSuccessor(movie);
    return new MovieWithSuccessor(movie, movieSuccessor);
  }

  private Movie getRandomMovieOrByTitle(String title) {
    Movie movie;
    if (title == null) {
      movie = movieService.getRecommendedMovie();
    } else {
      movie = movieService.getMovieByTitle(title);
    }
    return movie;
  }
}

File: outgoing-simple-kotlin-demo/src/main/kotlin/de/codecentric/cm4sb/demo/outgoing/endpoint/MovieController.kt
package de.codecentric.cm4sb.demo.outgoing.endpoint

import de.codecentric.cm4sb.demo.outgoing.domain.Movie
import de.codecentric.cm4sb.demo.outgoing.domain.MovieWithSuccessor
import de.codecentric.cm4sb.demo.outgoing.service.MovieSuccessorService
import de.codecentric.cm4sb.demo.outgoing.service.MovieService
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController


@RestController
class MovieController(val movieSuccessorService: MovieSuccessorService, val movieService: MovieService) {

    @GetMapping("/movies", produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getRecommendation(@RequestParam(required = false) title: String?): MovieWithSuccessor {
        val movie = getRandomMovieOrByTitle(title)

        val movieSuccessor = movieSuccessorService.getMovieSuccessor(movie)
        return MovieWithSuccessor(movie, movieSuccessor)
    }

    private fun getRandomMovieOrByTitle(title: String?): Movie {
        val movie = if (title == null) {
            movieService.getRecommendedMovie()
        } else {
            movieService.getMovieByTitle(title)
        }
        return movie
    }
}
